### ISSUE TITLE
glTF: Importing nodes with partly negative scale values -> all scale values are negative

### ISSUE THREAD

**Issue by ParadoxDaKo** [assimp/assimp#3784]
When decomposing a transformation matrix, [Assimp uses the trafo matrix's determinant to assign the sign of the scaling values](https://github.com/assimp/assimp/blob/master/include/assimp/matrix4x4.inl#L384):

```cpp
	/* extract the scaling factors */ \
	pScaling.x = vCols[0].Length(); \
	pScaling.y = vCols[1].Length(); \
	pScaling.z = vCols[2].Length(); \
	\
	/* and the sign of the scaling */ \
	if (Determinant() < 0) pScaling = -pScaling; \
```

Now, when we have glTF nodes with partly negative scale values such as this one:

```json
{
    "mesh" : 0,
    "name" : "Cube",
    "scale" : [
        -1,
        1,
        1
    ],
    "translation" : [
        0,
        -1,
        0
    ]
}
```
, then load a scene with this node with Assimp into our aiScene*, then try to read the scale values of this aiNode* via ```aiMatrix4x4t<TReal>::Decompose (aiVector3t<TReal>& pScaling, aiQuaterniont<TReal>& pRotation, aiVector3t<TReal>& pPosition)```, the scale vector is [-1, -1, -1].

Exporting a mesh node with this scale in Blender, then reimporting it in Blender gives me the scale values [-1, 1, 1].

What exactly is going wrong and where?

**Comment by JC3** [comment]
Just out of curiosity, what is the rotation in the transform decomposition for a scaled and unscaled node?

**Comment by ParadoxDaKo** [comment]
The rotation quarternions for both scaled and unscaled nodes are both [1, 0, 0, 0].

So this means Assimp applies the second decomposition out of the two you described? Ergo compensating flipped scale components in the rotation? 
If so, that's good to know - after applying the first decomposed rotation and scale values to my imported node manually, the results look equivalent, too.

... I assume there is no way to make Assimp use the first decomposition? 😅

**Comment by JC3** [comment]
> So this means that for nodes with negative scale values, Assimp applies the second decomposition out of the two you described? Ergo compensating flipped scale components in the rotation?

Yes, it appears to be producing a different decomposition, equivalent to but different than the one you expect.

> If so, that's good to know - after applying the first decomposed rotation and scale values to my imported node manually, the results look equivalent, too.

That's good confirmation. 

> ... I assume there is no way to make Assimp use the first decomposition? 😅

Unfortunately I'm not that familiar with assimp's internals but, my gut says no (I could be wrong though). What you're seeing is sort of (*sort of*) like the difference between *cos(π)* and *cos(-π)*, or *1²* and *(-1)²*, or *six* and a *half dozen*, or (*SORT OF*) the ambiguities you run into when trying to extract Euler angles from a system in gimbal lock. The decomposition you prefer sort of depends on your specific needs. The decomposition that the program produces just... is what it is. It's the decomposition you deserve, but maybe not the one you need right now. 😂

You can do it heuristically, though. For example, if it's important to you to minimize the number of negative scale values, then you could check to see if 2 or 3 of them are negative and, if so, flip the sign of two of the negative ones then rotate your quaternion to make up for it. Or whatever. Otoh, if all you're after is the end result, then it doesn't really matter.

**Comment by tellypresence** [comment]
@ParadoxDaKo Is this still a problem?

**Comment by tellypresence** [comment]
Closing due to no response to followup
